specs/phase1/beacon-chain.md

```diff
@@ -484,6 +485,10 @@ def process_shard_header(state: BeaconState,
     ))
 ```
 
+The length-and-degree proof works as follows. For a block B with length `l` (so `l` nonzero values in `[0...MAX_SAMPLES_PER_BLOCK-1]`), the length proof is supposed to be `(B / Z) * (r + X**(len(SETUP)-l))`, where `Z` is the minimal polynomial that is zero over `[l...MAX_SAMPLES_PER_BLOCK-1]` (see `SIZE_CHECK_POINTS` above). The goal is to ensure that a proof can only be constructed if (i) `B / Z` is itself non-fractional, meaning that `B` is a multiple of `Z`, and (ii) `deg(B) < MAX_SAMPLES_PER_BLOCK` (the block is not oversized).
+
+This is done by making the proof be a random linear combination of `B / Z` and `(B / Z) * (X**(len(SETUP)-l)`. The length proof will have the degree of `(B / Z) * X**(len(SETUP)-l)`, so `deg(B) - (MAX_SAMPLES_PER_BLOCK - l) + len(SETUP) - l`, simplified to `deg(B) - MAX_SAMPLES_PER_BLOCK + len(SETUP)`. Because it's only possible to generate proofs for polynomials with degree `< len(SETUP)`, it's this only possible to generate the proof if this expression is less than `len(SETUP)`, meaning that `deg(B)` must be strictly less than `MAX_SAMPLES_PER_BLOCK`.
```

Comment:
The current code does not implement the random linear combination. But I believe it's not necessary anyway: We can simply make the combined max degree length proof be `( B * X **(len(SETUP) - l) ) / Z`, which has a non-fractional solution only of `B` is `0` at the zeroes of `Z`.

Reply:
So on the idea of getting rid of the zero check alltogether, and making it a pure degree check: I've got some python code here that interpolates a polynomial from only evaluations on a part of a subgroup:
https://github.com/ethereum/research/blob/master/polynomial_reconstruction/interpolate_polynomial_without_zeroes.py
Excluding the part that can be precomputed, it takes about 4s, which I'd reckon can be brought down to 0.4s in a well optimized implementation. So it feels like that's possible and I would think we should just do it?
This will lead to some savings in the proof computation, so we may well end up even or with some small savings. But we get a much nicer code for length verification which I think is worth it.
Also a nice effect is that the samples become more effective (=better coding rate) for smaller blocks.

Reply:
[I implemented this as pure degree proofs for now, suggest we discuss the merits in this thread]